src/digraph-cyclicity-hadwick-james.jl

```diff
@@ -0,0 +1,97 @@
+@doc_str """
+Find circuits (including self-loops) using the algorithm of Hadwick & James
+("Enumerating Circuits and Loops in Graphs with Self-Arcs and Multiple-Arcs", 2008)
+"""
+function simplecycles_hadwick_james end
+@traitfn function simplecycles_hadwick_james(g::::IsDirected)
+    nvg = nv(g)
+    B = [Int[] for i in 1:nvg]
+    blocked = [false for i in 1:nvg]
+    stack = Int[]
+    cycles = Vector{Vector{Int}}()
+    for v in vertices(g)
+        circuit_recursive!(g, v, v, blocked, B, stack, cycles)
+        resetblocked!(blocked)
+        resetB!(B)
+    end
+    return cycles
+end
+
+"""
+Reset B work structure.
+"""
+resetB!(B) = map!(empty!, B, B)
+
+"""
+Reset vector of "blocked" vertices.
+"""
+resetblocked!(blocked) = fill!(blocked, false)
+
+@doc_str """
+Find circuits starting from v1 recursively.
+"""
+function circuit_recursive! end
+@traitfn function circuit_recursive!(g::::IsDirected, v1::Int, v2::Int, blocked::Vector{Bool}, B::Vector{Vector{Int}}, stack::Vector{Int}, cycles::Vector{Vector{Int}})
+    f = false
+    push!(stack, v2)
+    blocked[v2] = true
+
+    Av = out_neighbors(g, v2)
+    for w in Av
+        (w < v1) && continue
+        if w == v1 # Found a circuit
+            push!(cycles, copy(stack))
+            f = true
+        elseif !blocked[w]
+            if circuit_recursive!(g, v1, w, blocked, B, stack, cycles)
+                f = true
+            end
```

Comment:
can you move the `f = true` to before line 42 since it's being set in both branches?